src/main/java/com/uid2/admin/audit/QLDBAuditWriter.java

```diff
@@ -11,47 +11,79 @@
 import software.amazon.qldb.Result;
 import software.amazon.qldb.RetryPolicy;
 
-import java.util.ArrayList;
-import java.util.List;
+import java.util.*;
+import java.util.concurrent.atomic.AtomicBoolean;
 
 public class QLDBAuditWriter implements AuditWriter{
     private static final IonSystem ionSys = IonSystemBuilder.standard().build();
     private static final Logger logger = LoggerFactory.getLogger(QLDBAuditWriter.class);
     private static final Logger auditLogger = LoggerFactory.getLogger("com.uid2.admin.audit");
-    private final QldbDriver qldbDriver;
+    private QldbDriver qldbDriver;
     private final String logTable;
     private final boolean qldbLogging;
     public QLDBAuditWriter(JsonObject config){
-        qldbDriver = QldbDriver.builder()
-                .ledger(config.getString("qldb_ledger_name"))
-                .transactionRetryPolicy(RetryPolicy.builder().maxRetries(3).build())
-                .sessionClientBuilder(QldbSessionClient.builder())
-                .build();
+        try {
+            qldbDriver = QldbDriver.builder()
+                    .ledger(config.getString("qldb_ledger_name"))
+                    .transactionRetryPolicy(RetryPolicy.builder().maxRetries(3).build())
+                    .sessionClientBuilder(QldbSessionClient.builder())
+                    .build();
+        }
+        catch (Exception e){
+            logger.error("cannot establish connection with QLDB");
+        }
         logTable = config.getString("qldb_table_name");
         qldbLogging = config.getBoolean("enable_qldb_admin_logging");
     }
     @Override
-    public void writeLog(AuditModel model) {
-        if(model == null){ //should never be true, but check exists in case
-            return;
-        }
-        if(qldbLogging) {
-            qldbDriver.execute(txn -> {
-                List<IonValue> sanitizedInputs = new ArrayList<>();
-                JsonObject jsonObject = model.writeToJson();
-                StringBuilder query = new StringBuilder("UPDATE " + logTable + " AS t SET data = ?");
-                sanitizedInputs.add(ionSys.newLoader().load(jsonObject.toString()).get(0));
-                query.append(" WHERE t.data.itemType = ? AND t.data.itemKey = ?");
-                sanitizedInputs.add(ionSys.newString(jsonObject.getString("itemType")));
-                sanitizedInputs.add(ionSys.newString(jsonObject.getString("itemKey")));
-
-                Result r = txn.execute(query.toString(), sanitizedInputs);
-                if (!r.iterator().hasNext()) {
-                    logger.warn("Malformed audit log input: no log written to QLDB");
+    public boolean writeLogs(Collection<AuditModel> models) {
+        AtomicBoolean successfulLog = new AtomicBoolean(true);
```

Comment:
I don't see multi-threading here, why use atomic?

Reply:
Atomic here is just serving as a container for a boolean as anything referenced in the lambda expression must be "effectively final" or a compile error is thrown